### PR TITLE
Prepopulate data blocks to block cache during Flush

### PR DESCRIPTION
**Pull request by sagar0**
Prepopulate data blocks to block cache at Flush time. A new BlockBasedTable option, `prepopulate_data_blocks`, has been added to control this feature (default: turned off).

This piece is part of a larger feature which we have been calling "Cache Warming". 

This helps in improving the block cache hit rate leading to less IO incurred in reading back the flushed blocks from device, especially when using Direct IO, and thereby also reducing some tail latency (if it is due to flush/compaction block invalidations). 
Workloads which exhibit high temporal locality would especially benefit for it. 
Example: Before flush most of the recently written data is in mem-tables, and after flush some data gets written out to the device. If a lot of reads are still to the data that recently got flushed, your service would incur IO to bring those recently flushed blocks back into memory. With this feature, those blocks are already prepopulated to the block cache. 

Notes:
This could end up evicting blocks which are already at the tail end of the LRU block cache though. Also, this pre-population is done even before the new L0 file is added to the LSM manifest ... so there is a slight chance that the added blocks are invalid if the Flush fails/aborts (which rarely happens I believe)

For future:
- Index and filter blocks could also be pre-populated in the same way.
- Data blocks from higher levels which are already in memory could be pre-populated (replaced with newly created data blocks) during compactions as they would become invalidated at the end of a compaction.
- Maybe a background thread to go about evicting invalidated data blocks.

Test Plan:
Added a new test `DBBlockCacheTest. WarmCacheWithL0DataBlocks`.

### CODE REVIEW

File: table/block_based_table_builder.cc [hunk 1]
```diff
@@ -659,15 +681,50 @@ static void DeleteCachedBlockContents(const Slice& /*key*/, void* value) {
...
+      size_t charge = block->ApproximateMemoryUsage();
+
+      block_cache->Insert(key, block, charge, &DeleteCachedEntry<Block>);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_ADD);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_DATA_ADD);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_BYTES_WRITE, charge);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_DATA_BYTES_INSERT, charge);
+      return Status::OK();
+    }
+  }
```
Comment (siying) [inline table/block_based_table_builder.cc]: I'm confused a little bit. It feels that compress and uncompress cache are totally separated logic. Can we make them in two different functions, and call them accordingly?
I feel it is clearer than a if here and return in the end of the if. The branching can happen in the caller.

File: table/block_based_table_builder.cc [hunk 2]
```diff
@@ -659,15 +681,50 @@ static void DeleteCachedBlockContents(const Slice& /*key*/, void* value) {
...
+
+      Block* block =
+          new Block(std::move(results), kDisableGlobalSequenceNumber);
+      size_t charge = block->ApproximateMemoryUsage();
+
+      block_cache->Insert(key, block, charge, &DeleteCachedEntry<Block>);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_ADD);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_DATA_ADD);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_BYTES_WRITE, charge);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_DATA_BYTES_INSERT, charge);
```
Comment (siying) [inline table/block_based_table_builder.cc]: Is there some code we can consolidate with populating cache in the read path? A lot of these seem to be the same.

File: table/block_based_table_builder.cc [hunk 3]
```diff
@@ -659,15 +681,50 @@ static void DeleteCachedBlockContents(const Slice& /*key*/, void* value) {
...
                                                   const CompressionType type,
-                                                  const BlockHandle* handle) {
+                                                  const BlockHandle* handle,
+                                                  bool is_data_block) {
   Rep* r = rep_;
-  Cache* block_cache_compressed = r->table_options.block_cache_compressed.get();
 
+  // Uncompressed regular block cache
+  if (r->table_options.prepopulate_data_blocks && is_data_block) {
+    assert(r->level == 0);
```
Comment (riversand963) [inline table/block_based_table_builder.cc]: Seems this chunk of code will also be used when populating block cache with other types of blocks. Maybe put into a function?

**Comment by sagar0** [comment]
I need to address some comments and make some slight changes, before this can be reviewed again.

File: include/rocksdb/table.h [hunk 4]
```diff
@@ -257,6 +257,19 @@ struct BlockBasedTableOptions {
...
+  // On a flush, the data that is in memory (in memtables) get flushed to the
+  // device. If using Direct IO, additional IO is incurred to read this data
+  // back into memory again, which is avoided by enabling this option.
+  // This further helps if the workload exhibits high temporal locality, where
+  // most of the reads go to recently written data.
+  //
+  // TODO: Right now, this is enabled only for flush. We plan to expand this
+  // option to cover compactions in the future (as the blocks from old files are
+  // invalidated at the end of a compaction.)
+  bool prepopulate_data_blocks = false;
```
[the remark this answers is not in the document]
Comment (siying) [reply include/rocksdb/table.h]: This comment still remains unaddressed.

**Pull request closed without merging**